### PR TITLE
Add gradient clipping in the backward pass with a hook

### PR DESCRIPTION
**Pull request by matt-gardner**
This solves the exploding gradient problem.

### CODE REVIEW

**Comment by DeNeutoy** [comment]
This is so great. Doing this in Tensorflow would have been extremely tricky, I think. The `backward_hooks` is a really nice idea, and I can imagine doing other interesting stuff to them as well, e.g. variance reduction and stuff.

**Comment by matt-gardner** [comment]
I was curious about how hard it would be, too, so I spent a bit looking.  Looks like you [_can_](https://stackoverflow.com/questions/43839431/tensorflow-how-to-replace-or-modify-gradient) do it in tensorflow, though it looks like you would have to modify model code, and couldn't just do it in the `Trainer` like we can do with pytorch.  I'm quite sure you _couldn't_ do this with Keras, though.

**Comment by DeNeutoy** [comment]
Yeah, I think you're right. It's extremely frustrating to only find this now; thinking back, I'm almost certain this would have fixed the issues with the sentinel pointer network stuff I was doing.
